tests/test_device_validator.py

```diff
@@ -32,262 +33,230 @@
-        with self.assertRaises(D3ADeviceException):
-            validate_storage_device(energy_rate_decrease_per_update=-3)
-
-        self.assertIsNone(validate_storage_device(fit_to_limit=False,
-                                                  energy_rate_increase_per_update=3))
-        self.assertIsNone(validate_storage_device(fit_to_limit=False,
-                                                  energy_rate_decrease_per_update=3))
-        with self.assertRaises(D3ADeviceException):
-            validate_storage_device(fit_to_limit=True,
-                                    energy_rate_increase_per_update=3)
-        with self.assertRaises(D3ADeviceException):
-            validate_storage_device(fit_to_limit=True,
-                                    energy_rate_decrease_per_update=3)
-
-    def test_commercial_producer_setting(self):
-        self.assertIsNone(validate_commercial_producer(energy_rate=10))
-        with self.assertRaises(D3ADeviceException):
+VALID_LOAD_PROFILE = {"1": 3, "10": 4.2, "22:00": 0.1}
+
+
+class TestValidateDeviceSettings:
+    """Test the validation of the devices."""
+
+    @staticmethod
+    @pytest.mark.parametrize("valid_arguments", [
+        {"avg_power_W": 100},
+        {"hrs_per_day": 0},
+        {"hrs_per_day": 24},
+        {"final_buying_rate": 0},
+        {"initial_buying_rate": 0},
+        {"initial_buying_rate": 5, "final_buying_rate": 10},
+        {"hrs_of_day": [0, 3, 20, 21, 22, 23, 24]},
+        {"hrs_of_day": [1, 2, 3, 5, 6, 10], "hrs_per_day": 6},
+        {"energy_rate_increase_per_update": 0},
+        {"fit_to_limit": False, "energy_rate_increase_per_update": 2},
+        {"daily_load_profile": VALID_LOAD_PROFILE}
+    ])
+    def test_load_device_setting_succeeds(valid_arguments):
+        """The PV device validation succeeds when valid arguments are provided."""
+        assert validate_load_device(**valid_arguments) is None
+
+    @staticmethod
+    @pytest.mark.parametrize("failing_arguments", [
+        {"avg_power_W": -1},
+        {"hrs_per_day": 25},
+        {"final_buying_rate": -10},
+        {"initial_buying_rate": -2},
+        {"initial_buying_rate": 10, "final_buying_rate": 5},
+        {"hrs_of_day": [20, 21, 22, 23, 24, 25, 26]},
+        {"hrs_of_day": [1, 2, 3, 4, 5, 6], "hrs_per_day": 10},
+        {"energy_rate_increase_per_update": -1},
+        {"fit_to_limit": True, "energy_rate_increase_per_update": 2},
+        {"avg_power_W": 100, "daily_load_profile": VALID_LOAD_PROFILE},
+        {"hrs_per_day": 1, "daily_load_profile": VALID_LOAD_PROFILE},
+        {"hrs_of_day": [1, 2, 3, 4, 5, 6], "daily_load_profile": VALID_LOAD_PROFILE}
+    ])
+    def test_load_device_setting_fails(failing_arguments):
+        """The PV device validation fails when incompatible arguments are provided."""
+        with pytest.raises(D3ADeviceException):
+            validate_load_device(**failing_arguments)
+
+    @staticmethod
+    @pytest.mark.parametrize("valid_arguments", [
```

Comment:
This is a good opportunity to talk about it, up until now we used this package for parameterized tests https://pypi.org/project/parameterized/ (due to lack of the parametrize parameter in pytest.mark some years ago IIRC). Since such functionality is available in pytest now I think it is worth it to use it for all new code and gradually deprecate the use of the parameterized package (I am not sure whether pytest supports parameterized classes, but we do not use this feature anyways). Nothing to do in the context of this PR, just something to keep in mind for future PRs.

Reply:
I agree, I think we can gradually refactor existing code when we deal with it, until we can completely remove that dependency.